libr/anal/type.c

```diff
@@ -330,35 +279,81 @@ static RAnalBaseType *get_union_type(RAnal *anal, const char *sname) {
 	return NULL;
 }
 
+static RAnalBaseType *get_typedef_type(RAnal *anal, const char *sname) {
+	r_return_val_if_fail (anal && R_STR_ISNOTEMPTY (sname), NULL);
+
+	RAnalBaseType *base_type = r_anal_base_type_new (R_ANAL_BASE_TYPE_KIND_TYPEDEF);
+	if (!base_type) {
+		return NULL;
+	}
+
+	base_type->type = get_type_data (anal->sdb_types, "typedef", sname);
+	if (!base_type->type) {
+		goto error;
+	}
+	return base_type;
+
+error:
+	r_anal_base_type_free (base_type);
+	return NULL;
+}
+
+static RAnalBaseType *get_atomic_type(RAnal *anal, const char *sname) {
+	r_return_val_if_fail (anal && R_STR_ISNOTEMPTY (sname), NULL);
+
+	RAnalBaseType *base_type = r_anal_base_type_new (R_ANAL_BASE_TYPE_KIND_ATOMIC);
+	if (!base_type) {
+		return NULL;
+	}
+
+	base_type->type = get_type_data (anal->sdb_types, "type", sname);
+	if (!base_type->type) {
+		goto error;
+	}
+
+	RStrBuf key;
+	r_strbuf_init (&key);
+	r_strbuf_setf (&key, "type.%s.size", sname);
+	base_type->size = sdb_num_get (anal->sdb_types, r_strbuf_get (&key), 0);
+	r_strbuf_fini (&key);
+
+	return base_type;
+
+error:
+	r_anal_base_type_free (base_type);
+	return NULL;
+}
+
 // returns NULL if name is not found or any failure happened
 R_API RAnalBaseType *r_anal_get_base_type(RAnal *anal, const char *name) {
 	r_return_val_if_fail (anal && name, NULL);
 
 	char *sname = r_str_sanitize_sdb_key (name);
 	const char *type = sdb_const_get (anal->sdb_types, sname, NULL);
-
-	// Right now just types: struct, enum, union are supported
-	if (!type || !(strcmp (type, "enum") || strcmp (type, "struct") || strcmp (type, "union"))) {
+	if (!type) {
 		free (sname);
 		return NULL;
 	}
-	// Taking advantage that all 3 types start with distinct letter
-	// because the strcmp condition guarantees that only those will get to this flow
-	RAnalBaseType *base_type = NULL;
 
-	switch (type[0]) {
-	case 's': // struct
+	RAnalBaseType *base_type = NULL;
+	if (!strcmp (type, "struct")) {
 		base_type = get_struct_type (anal, sname);
-		break;
-	case 'e': // enum
+	} else if (!strcmp (type, "enum")) {
 		base_type = get_enum_type (anal, sname);
-		break;
-	case 'u': // union
+	} else if (!strcmp (type, "union")) {
 		base_type = get_union_type (anal, sname);
-		break;
+	} else if (!strcmp (type, "typedef")) {
```

Comment:
right now `struct C` is the same as `typedef struct {} C` in r2 types. both are accessiblel under the name "C", i assume this change wont affect the current C parser. but do you think it should be treated differently?

Reply:
I think the behavior that `struct C` must always be used as `struct C` instead of just `C` when it is not a typedef is very C-specific (C++ already doesn't enforce that) and not necessarily relevant for reverse engineering generic binaries written in any language.
Also, currently `typedef struct {} C` isn't directly representable in r2 because we can't have anonymous structs before maybe refactoring the types database out of sdb later.

Reply:
it is representable right now because it uses the typedefed name and the struct name is ignored. but it works as i said. you can test it. i did that on purpose for simplicity btw, so im aware of that behaviour and this is why im asking for feedback

Reply:
Yes, I think considering the current database design, this behavior is good.